src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1173 @@
+    #[cfg(not(feature = "with_ctap2_1"))]
+    let min_pin_length = 4;
+    if pin.len() < min_pin_length || pin.len() == PIN_PADDED_LENGTH {
+        // TODO(kaczmarczyck) check 4 code point minimum instead
+        return false;
+    }
+    let mut pin_hash = [0u8; 16];
+    pin_hash.copy_from_slice(&Sha256::hash(&pin[..])[..16]);
+    persistent_store.set_pin_hash(&pin_hash);
+    true
+}
+
+#[cfg(feature = "with_ctap2_1")]
+#[cfg_attr(test, derive(IntoEnumIterator))]
+// TODO remove when all variants are used
+#[allow(dead_code)]
+pub enum PinPermission {
+    // All variants should use integers with a single bit set.
+    MakeCredential = 0x01,
+    GetAssertion = 0x02,
+    CredentialManagement = 0x04,
+    BioEnrollment = 0x08,
+    PlatformConfiguration = 0x10,
+    AuthenticatorConfiguration = 0x20,
+}
+
+pub struct PinProtocolV1 {
+    key_agreement_key: crypto::ecdh::SecKey,
+    pin_uv_auth_token: [u8; PIN_TOKEN_LENGTH],
+    consecutive_pin_mismatches: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions_rp_id: Option<String>,
+}
+
+impl PinProtocolV1 {
+    pub fn new(rng: &mut impl Rng256) -> PinProtocolV1 {
+        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        let pin_uv_auth_token = rng.gen_uniform_u8x32();
+        PinProtocolV1 {
+            key_agreement_key,
+            pin_uv_auth_token,
+            consecutive_pin_mismatches: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        }
+    }
+
+    /// Decrypts the encrypted pin_hash and compares it to the stored pin_hash.
+    /// Resets or decreases the PIN retries, depending on success or failure.
+    fn verify_pin_hash_enc(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        aes_dec_key: &crypto::aes256::DecryptionKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        match persistent_store.pin_hash() {
+            Some(pin_hash) => {
+                if self.consecutive_pin_mismatches >= 3 {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                }
+                // We need to copy the pin hash, because decrementing the pin retries below mutably
+                // borrows from the same persistent_store reference, and therefore invalidates the
+                // pin_hash reference.
+                let pin_hash = pin_hash.clone();
+                persistent_store.decr_pin_retries();
+                if pin_hash_enc.len() != PIN_AUTH_LENGTH {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_INVALID);
+                }
+
+                let iv = [0u8; 16];
+                let mut blocks = [[0u8; 16]; 1];
+                blocks[0].copy_from_slice(&pin_hash_enc[0..PIN_AUTH_LENGTH]);
```

Comment:
If the length has already been checked, this indexing `0..PIN_AUTH_LENGTH` is unnecessary. Just make sure that there is a test where `pin_hash_enc` has too many (or too few) bytes.

Reply:
Done